hail/src/main/scala/is/hail/shuffler/Codec.scala

```diff
@@ -0,0 +1,28 @@
+package is.hail.shuffler
+
+import is.hail.annotations._
+import is.hail.expr.types.physical._
+import is.hail.io._
+import is.hail.utils._
+import java.io.{ ByteArrayOutputStream, ByteArrayInputStream }
+
+final class Codec (
+  spec: BufferSpec,
+  wireType: PType
```

Comment:
this feels conceptually wrong -- the wiretype is an EType, not a PType.

can this class wrap a `TypedCodecSpec` instead?

There's also no reason* why the input PType and output PType have to be the same.

`*` except our current 1:1 mapping between virtual and physical type :(

Reply:
Hmm. How do I get a (any) compatible `PType` from an `EType`?

Reply:
the best way is to use the TypedCodecSpec methods.

The buildDecoder method takes a requested virtual type (the virtual type of the input ptype) and returns a decoded ptype

Reply:
Or does the client need to specify both an etype and a ptype?